### ISSUE TITLE
Context without API

### ISSUE THREAD

**Issue by Haeri** [glfw/glfw#1789]
I was wondering if it is possible to use glfw without any backend api. The intent being the implementation of a software renderer, where all rendering happens on the CPU and the resulting pixel buffer is then provided to glfw to be presented in a window. I suspect it would be possible by creating a very simple OpenGL context that receives a buffer, but is there no other way, that does not require the presence of any backend api for example on embedded systems?

**Comment by GamesTrap** [comment]
Yes it's possible to create Windows without a context in GLFW just have a look in the docs:
https://www.glfw.org/docs/3.3/context_guide.html#context_less & https://www.glfw.org/docs/3.3/window_guide.html#GLFW_CLIENT_API_hint

**Comment by Haeri** [comment]
@GamesTrap Thanks for the response. I was aware that there is a GLFW_NO_API flag, as I am using that for my Vulkan implementation but I am still not sure, how to interface with the GLFW api to provide my bitmap.

Having this as a basis, where would I pass my bitmap buffer to draw in the window?

```c++
#include <GLFW/glfw3.h>

#include <stdlib.h>
#include <stdio.h>


static void error_callback(int error, const char* description)
{
    fprintf(stderr, "Error: %s\n", description);
}

static void key_callback(GLFWwindow* window, int key, int scancode, int action, int mods)
{
    if (key == GLFW_KEY_ESCAPE && action == GLFW_PRESS)
        glfwSetWindowShouldClose(window, GLFW_TRUE);
}

int main(void)
{
    const int width = 640;
    const int height = 480;

    GLFWwindow* window;

    glfwSetErrorCallback(error_callback);

    if (!glfwInit())
        exit(EXIT_FAILURE);

    glfwWindowHint(GLFW_CLIENT_API, GLFW_NO_API);

    window = glfwCreateWindow(width, height, "Simple example", NULL, NULL);
    if (!window)
    {
        glfwTerminate();
        exit(EXIT_FAILURE);
    }

    glfwSetKeyCallback(window, key_callback);

    while (!glfwWindowShouldClose(window))
    {
        glfwPollEvents();
    }

    glfwDestroyWindow(window);

    glfwTerminate();
    exit(EXIT_SUCCESS);
}

```

**Comment by elmindreda** [comment]
There is not yet any API in GLFW for putting an image buffer into the window without going through some rendering API like OpenGL or Vulkan.

**Comment by thedmd** [comment]
@Haeri [PixelToaster](https://github.com/thedmd/pixeltoaster) was created with software rasterization in mind. You can create a window and put buffer content on the screen. Input and Timer is also provided.

With GLFW you can achieve that by allowing it to create OpenGL context and update framebuffer context in each frame.

**Comment by aloucks** [comment]
GLFW has helper functions for creating a vulkan surface: 
https://www.glfw.org/docs/latest/vulkan_guide.html#vulkan_surface

Or you can use the native functions to acquire raw window primitives and use the Vulkan API directly:
https://www.glfw.org/docs/latest/group__native.html
https://vulkan-tutorial.com/Drawing_a_triangle/Presentation/Window_surface
